python/tskit/trees.py

```diff
@@ -4734,26 +4752,23 @@ def variants(
         # Only use impute_missing_data if isolated_as_missing has the default value
         if isolated_as_missing is None:
             isolated_as_missing = not impute_missing_data
+        if return_variant_copies is None:
+            return_variant_copies = True
+
         # See comments for the Variant type for discussion on why the
         # present form was chosen.
-        iterator = _tskit.VariantGenerator(
-            self._ll_tree_sequence,
+        variant = tskit.Variant(
+            self,
             samples=samples,
             isolated_as_missing=isolated_as_missing,
             alleles=alleles,
         )
-        for site_id, genotypes, alleles in iterator:
-            site = self.site(site_id)
-            if as_bytes:
-                if any(len(allele) > 1 for allele in alleles):
-                    raise ValueError(
-                        "as_bytes only supported for single-letter alleles"
-                    )
-                bytes_genotypes = np.empty(self.num_samples, dtype=np.uint8)
-                lookup = np.array([ord(a[0]) for a in alleles], dtype=np.uint8)
-                bytes_genotypes[:] = lookup[genotypes]
-                genotypes = bytes_genotypes.tobytes()
-            yield Variant(site, alleles, genotypes)
+        for site_id in range(self.num_sites):
+            variant.decode(site_id)
+            if return_variant_copies:
```

Comment:
Is it worth taking this condition out of the loop?
```
sites = range(self.num_sites)
if copy:
    for site_id in sites:
          yield variant.copy()
else:
     for site_id in sites:
          yield variant
```
Might be worth it since it's perf-critical?

Reply:
https://github.com/tskit-dev/tskit/pull/2172/commits/28a04ece4dce332aa3556fe3f08a8f8c433175fe